### ISSUE TITLE
$ref to definitions in another file not working?

### ISSUE THREAD

**Issue by mkaut** [json-editor/json-editor#923]
I cannot make references to external definitions to work, when running a copy of [https://json-editor.github.io/json-editor/](https://json-editor.github.io/json-editor/) on localhost (using python's http.server and with AJAX switched on).

To explain, let's say I have `common.schema.json` with

```javascript
{
  "$schema": "http://json-schema.org/draft-04/schema",
  "definitions": {
    "known-product": {
      "title": "product",
      "type": "string",
      "enum": [ "power", "hydrogen", "heat" ]
    }
  },
  "title": "test",
  "properties": {
    "customer": { "type": "string" },
    "product": { "$ref": "#/definitions/known-product" }
  }
}
```

and `test.schema.json` with

```javascript
{
  "$schema": "http://json-schema.org/draft-04/schema",
  "title": "Ref. tests",
  "properties": {
    "test-1": { "$ref": "common.schema.json#" },
    "test-2": { "$ref": "common.schema.json#/definitions/known-product" },
    "test-3": {
      "type": "object",
      "properties": {
        "Product": { "$ref": "./common.schema.json#/definitions/known-product" }
      }
    }
  }
}
```

Then, I put both schemas to the same folder as `index.html`, activate AJAX, and use `test.schema.json` in the editor.

In the generated editor, I would expect `test-1` to ignore `definitions` and show `customers` and `product`, while both `test-2` and `test-3` should read only `definitions/known-product` and therefore show only the product selection.
However, all three test show the same (customer and product), suggesting that that `$ref` ignores everything after `#` and simply includes the whole file.
Is this a bug, a missing feature, or am I just doing it wrong?

Note: it could be related to [this pull request](https://github.com/json-editor/json-editor/pull/284), but it is marked as "merged" and the examples all use internal references, despite the title mentioning "nonlocal definitions"...

**Comment by robocoder** [comment]
I'm encountering this issue too. I can see the remote schema being loaded and props being merged, but the schema itself isn't being updated, e.g.,
```
"test-1": { "$ref": "common.schema.json#" },
```
is replaced with
```
"test-1": { "$ref": "#/counter/1" },
```

**Comment by SebastienAndreo** [comment]
I think I have the same issue. 

here `animals.json` containing reusable definitions

```json
{
  "$schema": "http://json-schema.org/draft-04/schema",
  "$id": "animals.json#",
  "definitions": {
    "cat": {
      "title": "cat",
      "type": "object",
      "properties":{
        "name":{
          "type": "string"
        },
        "color":{
          "type": "string"
        }
      }
    },
    "dog": {
      "title": "dog",
      "type": "object",
      "properties":{
        "name":{
          "type": "string"
        },
        "size":{
          "type": "number"
        }
      }
    }
  }
}
```

And in animalsselector.json for the UI specification

```json
{
  "title": "Animals Selector",
  "$id": "animals_selector.json#",
  "type": "object",
  "format": "grid",
  "options": {
    "collapsed": false
  },
  "properties": {
    "category": {
      "title": "Animal Category",
      "type": "string",
      "enum": [
        "Cat",
        "Dog"
      ],
      "options": {
        "grid_columns": 12
      }
    },
    "cat": {
      "type": "object",
      "$ref": "animals.json#/definitions/cat",
      "options": {
        "dependencies": {
          "category": "Cat"
        },
        "grid_columns": 12
      }
    },
    "dog": {
      "type": "object",
      "$ref": "animals.json#/definitions/dog",
      "options": {
        "dependencies": {
          "category": "Dog"
        },
        "grid_columns": 12
      }
    }
  }
}
```
The UI works well if the cat and dog definitions are `animalsselector.json` if not the properties name, size ... do not appears.

**Comment by robocoder** [comment]
Yeah, when I dumped the refs, I found the key was malformed which is why the schemaloader couldn't resolve the JSON pointer path, eg "animals.json#/definitions/cat#/definitions/cat".

**Comment by schmunk42** [comment]
@mkaut and @SebastienAndreo could you try https://github.com/json-editor/json-editor/pull/928 and report if it fixes your problem.

**Comment by mkaut** [comment]
> @mkaut and @SebastienAndreo could you try #928 and report if it fixes your problem.

I apologize for a possibly stupid question, but how can I try it? I know very little about javascript and nothing about javascript development - I do not have npm and simply use json-editor from the provided cdn. Hence,I do not know how I can try a version that, as far as I can see, is not posted anywhere .. sorry.

**Comment by SebastienAndreo** [comment]
The problem solved

**Comment by mkaut** [comment]
I am sorry, but is it supposed to work in v.2.5.3? As far as I can see, v.2.5.3. was created 20 day ago, while the fixed was merged into master 23 days ago, i.e., 3 days before that.
I ask because neither mine or Sebastien's example work for me with v.2.5.3, I cannot figure out why...

**Comment by robocoder** [comment]
,Did you see the editor option for `ajax: true`?

**Comment by mkaut** [comment]
@robocoder: Yes, I see that the files are being loaded.
I have uploaded my setup I use [here](https://www.dropbox.com/s/nl5g2ldh8vlscie/json-editor_issue.zip?dl=0), including a python script that starts the python http-server I use for viewing the file.
Once `editor_test.html` (which is a simplified version of the [official demo](https://json-editor.github.io/json-editor/)) is opened, I select AJAX in the options.

When I then paste the content of `test.schema.json` into the Schema dialog and click on `Update Schema`, the generated Editor has 'customer' field in all three tests, suggesting that the `$ref` links the content of the linked file instead of the definitions-section.

When I instead paste the content of `animalselector.json`, the `cat` and `dog` objects have no properties, again suggesting that the definitions from the linked files were not read.

Do you get the same results (with my files)?
And if so, is there something wrong with the setup?

**Comment by robocoder** [comment]
You need to enable json-editor's `ajax` option.

In your `editor_test.html`, change this:
```
  var defaultOptions = {
    iconlib: 'fontawesome5',
    object_layout: 'normal',
    schema: defaultSchema,
    show_errors: 'interaction',
    theme: 'bootstrap4'
  }
```

to:
```
  var defaultOptions = {
    iconlib: 'fontawesome5',
    object_layout: 'normal',
    schema: defaultSchema,
    show_errors: 'interaction',
    theme: 'bootstrap4',
    ajax: true
  }
```

**Comment by mkaut** [comment]
@robocoder 
> You need to enable json-editor's `ajax` option.

Again, I have Ajax working, otherwise it would not read the linked json files at all. I tried adding the option just in case, but it did not change anything.
I have also made a simpler [HTML file](https://www.dropbox.com/s/pa938bbtsk7muil/test-input.html?dl=0), hard-coded to use the `test.schema.json` schema, with references to `common.schema.json` - but it still behaves the same way, i.e., all three tests include the 'customer' field (which should only exist in 'test-1'). :-(

**Comment by mkaut** [comment]
@robocoder 
I think there must be something wrong with my setup, though I cannot find out what. For example, this schema:
```
{
	"$schema": "http://json-schema.org/draft-04/schema",
	"title": "Model input",
	"type": "object",
	"required": ["parameters"],
	"properties": {
		"parameters": {
			"title": "Model parameters",
			"type": "object",
			"properties": {
				"currency_unit": {
					"title": "currency unit",
					"type": "string",
					"enum": ["€", "$", "£"],
					"default": "€"
				},
				"disc_rate": {
					"title": "discount rate [%]",
					"type": "number",
					"minimum": 0,
					"maximum": 10,
					"default": 4
				},
				"time_unit": {
					"title": "time unit",
					"type": "string",
					"enum": ["second", "min", "hour"],
					"default": "hour"
				}
			},
			"required": ["currency_unit"],
			"defaultProperties": ["currency_unit", "disc_rate"]
		}
	}
}
```
works as expected in the [json-editor demo](https://json-editor.github.io/json-editor/), i.e., displays the first two properties by default. It also keeps displaying them when I activate `display_required_only` option, showing that `defaultProperties` overrides `required`.

However, when I use the same schema locally, with [this html file](https://github.com/json-editor/json-editor/files/6285148/jsed-test.zip), then it always shows only the first (required) property:
![image](https://user-images.githubusercontent.com/9974117/114169552-010d5e80-9932-11eb-87dc-7f77a4dbfa2a.png)
It looks like it ignores both the `defaultProperties` value and the `display_required_only` option (which is at `false` by default).

I tested this in Firefox and Edge (on Windows) and both show the same.
I must admit I have run out of ideas what to test - but the point is that since I get this weird behaviour with my setup, then the issue reported above is just as likely to be my local problem as well :-(

**Comment by robocoder** [comment]
Try clearing your browser cache or reloading the test page while the devtools pane is open.

**Comment by mkaut** [comment]
Thanks again - I have now used Chrome (which I normally do not use), cleared its cache and for a good measure also disabled cache in the devtools pane.
This fixed the latest issue, but the original issue still remains: when I open [test-input.html](https://www.dropbox.com/s/pa938bbtsk7muil/test-input.html?dl=0), which has `$ref`s to [common.schema.json](https://www.dropbox.com/s/agzthskeagwi6xe/common.schema.json?dl=0), it still shows a 'customer' field in all three tests :-(

In other words, I do not see any difference compared to my original post .. do you get correct results with either my or Sebastien's example, using v.2.5.3 from the CDN?

**Comment by schmunk42** [comment]
@mkaut might be worth opening a new issue, ideally with a test.

**Comment by robocoder** [comment]
@mkaut sorry, I've been a bit busy to troubleshoot but I might have narrowed it down to `schemaloader.js`'s `_isLocalUrl()`. If you change your refs to have a leading slash, e.g.,`/common.schema.json`, the results are different.

**Comment by mkaut** [comment]
@robocoder Thanks, this fixed it! I can confirm that with a leading slash added to the `$ref`s, now both the examples work as expected. :-)
Is this a work-around for a bug in `schemaloader.js` (which will be fixed at some point), or is it meant to work like this? I ask because I do not see any leading slashes in [json-schema documentation](https://json-schema.org/understanding-json-schema/structuring.html), so if json-editor needs them, it should be mentioned in its documentation...

**Comment by robocoder** [comment]
I think this is a bug.

The current behaviour of `_isLocalUrl()` dates back to 348f4dd63c87c50bc4603de0698c439cf90a2870 but looking at the commit, the unit test doesn't support making this change, i.e., it doesn't test for the leading slash.

**Comment by schmunk42** [comment]
PRs welcome

**Comment by robocoder** [comment]
Ok, I understand what's going on. It was intended to handle the recursive case, e.g.,

initial schema:
```
{
  type: 'object',
  properties: {
    fname: { $ref: '/schema/main.json' },
    lname: { $ref: '/schema/main.json' }
  }
}
```

main.json
```
{
  $ref: 'registry/sub.json'
}
```

so this last $ref would resolve as `/schema/registry/sub.json` instead of `registry/sub.json'. The spec appears to be ambiguous which explains why https://github.com/Julian/jsonschema allows a custom Reference Resolver.

**Comment by robocoder** [comment]
p.s. I have a fix; I'll submit a PR as soon as I have time to write the tests for it
